internal/services/ingest.go

```diff
@@ -72,37 +80,64 @@ func NewIngestService(
 }
 
 func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger uint32) error {
-	heartbeat := make(chan any)
-	go trackServiceHealth(heartbeat, m.appTracker)
+	rpcHeartbeat := make(chan entities.RPCGetHealthResult, 1)
+	ingestHeartbeat := make(chan any, 1)
+
+	// Start service health trackers
+	go trackRPCServiceHealth(ctx, rpcHeartbeat, m.appTracker, m.rpcService)
+	go trackIngestServiceHealth(ctx, ingestHeartbeat, m.appTracker)
+
 	if startLedger == 0 {
 		var err error
 		startLedger, err = m.models.Payments.GetLatestLedgerSynced(ctx, m.ledgerCursorName)
 		if err != nil {
 			return fmt.Errorf("erorr getting start ledger: %w", err)
 		}
 	}
+
 	ingestLedger := startLedger
-	for ; endLedger == 0 || ingestLedger <= endLedger; ingestLedger++ {
-		time.Sleep(10 * time.Second)
-		ledgerTransactions, err := m.GetLedgerTransactions(int64(ingestLedger))
-		if err != nil {
-			log.Error("getTransactions: %w", err)
-			continue
-		}
-		heartbeat <- true
-		err = m.ingestPayments(ctx, ledgerTransactions)
-		if err != nil {
-			return fmt.Errorf("error ingesting payments: %w", err)
-		}
+	for endLedger == 0 || ingestLedger <= endLedger {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancelled: %w", ctx.Err())
+		case resp := <-rpcHeartbeat:
+			switch {
+			// Case-1: wallet-backend is running behind rpc's oldest ledger. In this case, we start
+			// ingestion from rpc's oldest ledger.
+			case ingestLedger < resp.OldestLedger:
+				ingestLedger = resp.OldestLedger
```

Comment:
Is this the desired functionality? Wouldn't a gap in the wallet backend's history be created if it was behind RPC's oldest ledger?

Reply:
Resolving this as we discussed in 1-1